File: Data Structures & Algorithms/balanced-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
  /**
   * @param {TreeNode} root
   * @return {boolean}
   */
  isBalanced(root) {
    if (root === null) return true;
    return (
      Math.abs(getHeight(root.left) - getHeight(root.right)) <= 1 &&
      this.isBalanced(root.left) &&
      this.isBalanced(root.right)
    );
  }
}

function getHeight(root) {
  if (root === null) return 0;
  return Math.max(getHeight(root.left), getHeight(root.right)) + 1;
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        count = 0

        def dfs(root: TreeNode, max_in_path: int) -> None:
            nonlocal count
            if root.val >= max_in_path:
                count += 1

            if root.left:
                dfs(root.left, max(max_in_path, root.val))
            if root.right:
                dfs(root.right, max(max_in_path, root.val))

        dfs(root, root.val)
        return count